lib/rules/media-feature-range-notation/index.mjs

```diff
@@ -80,20 +80,26 @@ const rule = (primary, _secondaryOptions, context) => {
 									-1,
 									{ value: unprefixedMediaFeature },
 								]),
-								node.name.before,
-								node.name.after.length > 0
-									? node.name.after
+								entry.node.name.before,
+								entry.node.name.after.length > 0
+									? entry.node.name.after
 									: [[TokenType.Whitespace, ' ', -1, -1, undefined]],
 							),
 							[operator, [TokenType.Delim, '=', -1, -1, { value: '=' }]],
-							node.value,
+							entry.node.value,
 						);
 
-						return;
-					}
+						const expectedMediaQueryList = mediaQueryList.map((mq) => mq.toString()).join(',');
```

Comment:
[ask] This PR increases `mq.toString()` in the loop, but I'm not sure about the effects on performance.

@romainmenke Can you please let us know if you have any concerns? 🙏🏼

Reply:
Serialization for these objects isn't fast, also not slow.

It's a bunch of recursive calls to `foo.toString()` where eventually tokens are found which each have string representations. All those bits of strings are concatenated.

So it's slow-ish because it can be a lot of function calls.
But it's fast-ish because the operations are always simple.

------

Does this change really increase the number of calls?
I thought it was only moved around a bit 🤔

Reply:
Thanks for the answer. It sounds like no significant issue. 👍🏼 

> Does this change really increase the number of calls?

If I understand correctly, this PR moves `mediaQueryList.map((mediaQuery) => mediaQuery.toString())` from outside to inside the `mediaQueryList.forEach()` loop. So, the fix count increases by the loop count.

https://github.com/stylelint/stylelint/blob/d1d11f2c9556b9274e900188df1416a161ba9ddf/lib/rules/media-feature-range-notation/index.mjs#L45

But if the loop count is a few, this may not be a problem. 😅

Reply:
Oh, wow, this was completely invisible to me in the git diff 😅 

------

https://github.com/stylelint/stylelint/blob/d1d11f2c9556b9274e900188df1416a161ba9ddf/lib/rules/media-feature-range-notation/index.mjs#L45

vs. 

https://github.com/stylelint/stylelint/blob/main/lib/rules/media-feature-range-notation/index.mjs

The larger issue is that multiple reports are possible. If this is a complex query it can contain multiple warnings.

This rule assumes that fixes and warnings are separate branches and that there is no overlap.

I need to think on how we can handle this better.
But we shouldn't merge this as is.

I will also try to formulate a failing test.

Reply:
> But we shouldn't merge this as is.
> I will also try to formulate a failing test.

Please do.

Reply:
```js
{
	code: `@media (min-width: 1px), (max-width: 2px) and (min-height: 3px) {}`,
	fixed: `@media (width >= 1px), (width <= 2px) and (height >= 3px) {}`,
	description: 'two range type media features in prefix notation',
	warnings: [
		{ message: messages.expected('context'), line: 1, column: 8, endLine: 1, endColumn: 24 },
		{ message: messages.expected('context'), line: 1, column: 26, endLine: 1, endColumn: 42 },
		{ message: messages.expected('context'), line: 1, column: 47, endLine: 1, endColumn: 64 },
	],
},
```

Before we started this effort the fix and warning execution branches were completely separate and fixes could do anything they needed to fix the source.

The final goal however is something somewhat distant from the current state.
We want fixes to be independent operations that do not have side effects or bleed.

In the above example we do have side effects and bleed.

After the first fix the AST will have changed and serialized it will look like:

```css
(width >= 1px), (max-width: 2px) and (min-height: 3px)
```

So the second fix will start from that point and won't be independent.
It will produce:

```css
(width >= 1px), (width <= 2px) and (min-height: 3px)
```

While it should instead produce:

```css
(min-width: 1px), (width <= 2px) and (min-height: 3px)
```

This however is not observable at this time because we don't yet return ranges or replacement texts. The end result will not be wrong. We will have a few too many mutations and serializations, but we will never have a hard bug.

I will open a separate issue for this because I don't want to block the progress here while I still want to formulate general guidance for this challenge. I think it will be most prominent with the resolved/flattened selectors for nesting.

Reply:
@romainmenke Thank you for the clear description with the actual example. I now realize this problem that you're concerned about, and I agree that we can discuss it in the follow-up issue. 👍🏼

Reply:
> we don't yet return ranges or replacement texts

This means that the way we will test will require to test only the range returned in correlation with the text.
i.e.
```
[
//...
  {
    range: { start: { line: 1, column: 27 }, end: { line: 1, column: 39 } },
    text: 'width <= 2',
  }
/...
]
```